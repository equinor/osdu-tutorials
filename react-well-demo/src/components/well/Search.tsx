import React, { ChangeEvent, useState, FormEvent, useEffect } from 'react';
import { Spin, Alert } from "antd";
import { LoadingOutlined } from "@ant-design/icons";
import './styles.css';
import { FoundWell } from './FoundWell';
import {useDispatch, useSelector} from "react-redux";
import {AppState} from "../../store";
import {findWellsByNameAction} from "../../store/well/actions";
import {loadWellboreTrajectoryAction} from "../../store/wellbore_trajectory/actions";

const noSearchHint = 'Results will be displayed here';
const noDataHint = 'No wells found';

export function Search() {
    const dispatch = useDispatch();

    const storedSearchName = useSelector((state: AppState) => state.wellSearch.searchName);
    const areWellsSearching = useSelector((state: AppState) => state.wellSearch.areWellsSearching);
    const areWellsSearched = useSelector((state: AppState) => state.wellSearch.areWellsSearched);
    const foundWells = useSelector((state: AppState) => state.wellSearch.foundWells);
    const searchError = useSelector((state: AppState) => state.wellSearch.searchError);

    const [searchName, setSearchName] = useState(storedSearchName);
    const loadingIcon = <LoadingOutlined style={{ fontSize: 24 }} spin/>;

    const handleSearchChange = (event: ChangeEvent<HTMLInputElement>) => {
        // note, that this value does not sync with the store
        // store will be updated only on an actual search action
        setSearchName(event.target.value);
    };

    const handleSubmit = (event: FormEvent | MouseEvent) => {
        event.preventDefault();
        dispatch(findWellsByNameAction(searchName));
    };

    useEffect(() => {
        // without this update, value from the store used only once.
        // and never realy watched, creating an illusion of a coherent behavior
        setSearchName(storedSearchName);
    }, [storedSearchName]);

    return (
        <div className="search">
            {/* a search form at the top */}
            <form className="search__area" onSubmit={handleSubmit}>
                <input
                    type="text"
                    className="search__text"
                    placeholder="Enter well name"
                    onChange={handleSearchChange}
                    value={searchName}
                />
                <input className="search__submit" type="submit" value="Search" onClick={handleSubmit} />
            </form>
            {/* a result representing area right under it */}
            <div className="search__well-area">
                {areWellsSearching ? (
                    // a progress, an anticipation
                    <Spin indicator={loadingIcon}/>
                ) : searchError ? (
                    // an error will be drawn here
                    <Alert
                        message="Cannot load well"
                        showIcon
                        description={String(searchError)}
                        type="error"
                    />
                ) : areWellsSearched ? (
                    // Successful search, two options now:
                    foundWells.length === 0 ? (
                        // nothing really found
                        // an honest descriptive message is much better that a blank space
                        <Alert message={noDataHint} type="warning" />
                    ) : (
                        // success, wells will be drawn
                        foundWells.map(well => (
                            <FoundWell
                                key={well.resourceId}
                                well={well}
                            />
                        ))
                    )
                ) : (
                    <Alert message={noSearchHint} type="success"/>
                )}
            </div>
        </div>
    );
}
