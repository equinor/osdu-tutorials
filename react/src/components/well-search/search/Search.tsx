import React, { ChangeEvent, useState, FormEvent, useEffect } from 'react';
import './styles.css';
import { Loader, Hint } from 'components/shared';
import { FoundWell } from '../found-well';
import { WellSearchResponse } from 'store/well-search';

export interface SearchProps {
  /**
   * a search value, stored in a redux state
   * though we will not update in on every input change,
   * we will keep an eye on it and make sure to use it if updated
   */
  storedSearchName: string;

  /** its text will be displayed instead of results list */
  searchError?: Error;

  /** flag to toggle an amazing loader */
  areWellsSearching: boolean;

  /**
   * with this knowledge we have an ability to inform a user that
   * the search itself was successful, just fruitless :)
   */
  areWellsSearched: boolean;

  /** we'll draw a cool list out of them */
  foundWells: WellSearchResponse[];

  /** search form submit */
  onSubmit: (name: string) => void;
}

const noSearchHint = 'Results will be displayed here';
const noDataHint = 'No wells found';
const cannotFindWellsMsg = 'Cannot find wells';

export function Search({
  onSubmit,
  storedSearchName,
  searchError,
  areWellsSearching,
  areWellsSearched,
  foundWells,
}: SearchProps) {
  const [searchName, setSearchName] = useState(storedSearchName);

  const handleSearchChange = (event: ChangeEvent<HTMLInputElement>) => {
    // note, that this value does not sync with the store
    // store will be updated only on an actual search action
    setSearchName(event.target.value);
  };

  const handleSubmit = (event: FormEvent | MouseEvent) => {
    event.preventDefault();
    onSubmit(searchName);
  };

  useEffect(() => {
    // without this update, value from the store used only once.
    // and never realy watched, creating an illusion of a coherent behavior
    setSearchName(storedSearchName);
  }, [storedSearchName]);

  return (
    <div className="search">
      {/* a search form at the top */}
      <form className="search__area" onSubmit={handleSubmit}>
        <input
          type="text"
          className="search__text"
          placeholder="Enter well name"
          onChange={handleSearchChange}
          value={searchName}
        />
        <input className="search__submit" type="submit" value="Search" onClick={handleSubmit} />
      </form>
      {/* a result representing area right under it */}
      <div className="search__well-area">
        {areWellsSearching ? (
          // a progress, an anticipation
          <Loader />
        ) : searchError ? (
          // an error will be drawn here
          <Hint title={cannotFindWellsMsg} subTitle={String(searchError)} />
        ) : areWellsSearched ? (
          // Successful search, two options now:
          foundWells.length === 0 ? (
            // nothing really found
            // an honest descriptive message is much better that a blank space
            <Hint subTitle={noDataHint} />
          ) : (
              // success, wells will be drawn
              foundWells.map(well => (
                <FoundWell
                  key={well.resourceId}
                  well={well}
                />
              ))
            )
        ) : (
                // it's a very beginning: nothing done right, nothing went wrong
                // a perfect time to give a hint ';..;'
                <Hint subTitle={noSearchHint} />
              )}
      </div>
    </div>
  );
}
