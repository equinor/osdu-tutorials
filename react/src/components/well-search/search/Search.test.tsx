import React from 'react';
import { shallow } from 'enzyme';
import { Search, SearchProps } from './Search';

describe('Search', () => {
  it('should render self and subcomponents', () => {
    const props: SearchProps = {
      onSubmit: jest.fn(),
      storedSearchName: 'storeSearchName',
      onLoadWellbores: jest.fn(),
      onFetchTrajectory: jest.fn(),
      onUnselectTrajectory: jest.fn(),
      searchError: undefined,
      areWellsSearching: false,
      areWellsSearched: false,
      foundWells: [],
      selectedWell: '',
      selectedTrajectories: [],
    };
    const wrapper = shallow(<Search {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
