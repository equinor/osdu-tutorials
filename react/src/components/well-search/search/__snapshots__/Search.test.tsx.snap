// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render self and subcomponents 1`] = `
<div
  className="search"
>
  <form
    className="search__area"
    onSubmit={[Function]}
  >
    <input
      className="search__text"
      onChange={[Function]}
      placeholder="Enter well name"
      type="text"
      value="storeSearchName"
    />
    <input
      className="search__submit"
      onClick={[Function]}
      type="submit"
      value="Search"
    />
  </form>
  <div
    className="search__well-area"
  >
    <Hint
      subTitle="Results will be displayed here"
    />
  </div>
</div>
`;
